dkg/bcast/server.go

```diff
@@ -48,32 +49,41 @@ type dedupKey struct {
 	MsgID  string
 }
 
+type messageIDFuncs struct {
+	callback     Callback
+	checkMessage CheckMessage
+}
+
 // server is a reliable-broadcast server.
 type server struct {
-	callbacksMutex sync.Mutex
-	callbacks      map[string]Callback
+	msgIDFuncsMutex sync.Mutex
+	msgIDFuncs      map[string]messageIDFuncs
 
+	hashFunc   hashFunc
 	signFunc   signFunc
 	verifyFunc verifyFunc
 
 	mu    sync.Mutex
 	dedup map[dedupKey][]byte // map[dedupKey]hash
 }
 
-func (s *server) getCallback(msgID string) (Callback, bool) {
-	s.callbacksMutex.Lock()
-	defer s.callbacksMutex.Unlock()
+func (s *server) getMessageIDFunc(msgID string) (messageIDFuncs, bool) {
+	s.msgIDFuncsMutex.Lock()
+	defer s.msgIDFuncsMutex.Unlock()
 
-	fn, found := s.callbacks[msgID]
+	fn, found := s.msgIDFuncs[msgID]
 
 	return fn, found
 }
 
-func (s *server) registerCallback(msgID string, cb Callback) {
-	s.callbacksMutex.Lock()
-	defer s.callbacksMutex.Unlock()
+func (s *server) registerMessageIDFuncs(msgID string, cb Callback, cm CheckMessage) {
+	s.msgIDFuncsMutex.Lock()
```

Comment:
I am wondering why don't we use `RWLock`s?

Reply:
Considering this mutex is only used at initialization phase, and that is essentially always only read, I don't see a point to use RWMutex here.

Generally though, I agree we should check whether we can use RWMutex widely.